### PR TITLE
task5 

### PR DESCRIPTION
**Pull request by abhin2002**

### CODE REVIEW

File: task_5/task5.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import os
+import urllib.request
+import pandas as pd
+import urllib.error
+
+class Downloader:
+    def __init__(self, pq_file: str):
+        self.df = pd.read_parquet(pq_file)
```
Comment (FayazRahman) [inline task_5/task5.py]: Don't read full file, specify `columns` to save memory and time

**Review by FayazRahman — CHANGES_REQUESTED** [review]

File: task_5/task5.py (hunk 1 above)
Comment (abhinpt0) [reply task_5/task5.py]: Done 👍

File: task_4/task4.py [hunk 2]
```diff
@@ -1,34 +1,34 @@
...
+    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3'
+}
+
+for i, url in enumerate(tqdm(urls, desc='Downloading images')):
+    try:
+        response = requests.get(url, headers=headers, stream=True)
+        response.raise_for_status()
+
+        # Save the image to a file
+        with open(f'images/image_{i}.jpg', 'wb') as f:
```
Comment (FayazRahman) [inline task_4/task4.py]: All files may not be `.jpg`

File: task_4/task4.py [hunk 3]
```diff
@@ -1,34 +1,34 @@
...
-        print(f'Failed to download image {i} from URL: {url}')
-        print(f'Error: {e}')
-
+import os
+import pyarrow.parquet as pq
+import requests
+from tqdm import tqdm
+
+# Load the parquet file
+table = pq.read_table('links.parquet')
```
Comment (FayazRahman) [inline task_4/task4.py]: Read required columns only.

File: task_5/task5.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        try:
+            urllib.request.urlretrieve(url, local_path)
+        except Exception as e:
+            return f"An error occurred while downloading the image: {e}"
+        return local_path
+
+    def __getitem__(self, key):
+        if isinstance(key, int):
+            url = self.df.iloc[key]['URL']
+            local_path = self.base_path + f'image_{key}.jpg'
```
Comment (FayazRahman) [inline task_5/task5.py]: All paths may not be `.jpg`

**Review by FayazRahman — CHANGES_REQUESTED** [review]

File: task_4/task4.py (hunk 3 above)
Comment (abhinpt0) [reply task_4/task4.py]: done

File: task_4/task4.py (hunk 2 above)
Comment (abhinpt0) [reply task_4/task4.py]: updated

**Review by FayazRahman — APPROVED** [review]

**Pull request merged**